### PR TITLE
Allow running on plain (non-git) directories

### LINKED ISSUE #71: Improve error when not run in a Git repository
```bash
$ mkdir /tmp/test ; cd /tmp/test ; darker .
error: Could not access 'HEAD'
Traceback (most recent call last):
  File "/home/akaihola/.virtualenvs/darker/bin/darker", line 33, in <module>
    sys.exit(load_entry_point('darker', 'console_scripts', 'darker')())
  File "/home/akaihola/prg/darker/src/darker/__main__.py", line 230, in main
    for path, old_content, new_content, new_lines in format_edited_parts(
  File "/home/akaihola/prg/darker/src/darker/__main__.py", line 64, in format_edited_parts
    changed_files = git_get_modified_files(srcs, revrange, git_root)
  File "/home/akaihola/prg/darker/src/darker/git.py", line 174, in git_get_modified_files
    lines = _git_check_output_lines(diff_cmd, cwd)
  File "/home/akaihola/prg/darker/src/darker/git.py", line 130, in _git_check_output_lines
    return check_output(cmd, cwd=str(cwd)).decode("utf-8").splitlines()
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'diff', '--name-only', '--relative', 'HEAD', '--', '.']' returned non-zero exit status 1.
```

### PR DESCRIPTION
**Pull request by ivanov**
I started off writing some code for better messages when running `darker` on a non-git repo, to fix #71, but then ended up writing enough to make it work anyway.

With this PR, you can `mkdir /tmp/test ; cd /tmp/test ; darker .` and no compaints are raised. Same goes for

```diff
$ mkdir -p /tmp/test ; cd /tmp/test ; echo "print   ('hello'  )" > a.py; darker --diff .;
--- a.py
+++ a.py
@@ -1 +1 @@
-print   ('hello'  )
+print("hello")
```

Opening this up for discussion to see if this if @akaihola would be amenable to something like this, before I start adding tests. In particular,  this will conflict with #138, though I also ran into AST verification failing when using `--isort`, my fix is slightly different (perhaps wrong?). 

#139 also came out of the work on this, but I decided to submit that one separately, since it's an independent change.

### CODE REVIEW

File: src/darker/__main__.py [hunk 1]
```diff
@@ -219,7 +230,16 @@ def main(argv: List[str] = None) -> int:
...
+    try:
+        changed_files = git_get_modified_files(paths, revrange, git_root)
+    except NotGitRespository:
+        changed_files = set()
+        for path in paths:
+            if str(path).endswith(".py"):
+                changed_files.add(Path(path))
+            else:
+                more_files = glob.glob(str(path) + "/**/*.py", recursive=True)
+                changed_files.update({Path(p) for p in more_files})
```
Comment (ivanov) [inline src/darker/__main__.py:242]: We can probably throw this logic into a separate `get_relevant_files_without_git` function, and keep the logic as 

```python
     try:
        changed_files = git_get_modified_files(paths, revrange, git_root)
    except NotGitRespository:
        changed_files = get_relevant_files_without_git(paths)
```

open to other names, too? Maybe `nogit_get_relevant_files`?

File: src/darker/__main__.py [hunk 2]
```diff
@@ -74,6 +80,11 @@ def format_edited_parts(
             )
             if enable_isort and not edited_linenums and edited == worktree_content:
                 logger.debug("No changes in %s after isort", src)
+                last_successful_reformat = (
+                    src,
+                    worktree_content,
+                    TextDocument.from_lines(""),
+                )
```
Comment (ivanov) [inline src/darker/__main__.py:87]: this was my fix for the false AST non-equivalence thing described in #137, though I make the `chosen` lines empty.

**Comment by akaihola** [comment]
Thanks @ivanov for working on this proposal.

I'm just wondering what the use case for Darker in a non-version-controlled source code tree is. You could just as well run Black, isort and any linters directly on the whole source tree since there are no known "modifications" to limit output to. Or am I misunderstanding your idea?

Is this just for convenience so you can use the same toolchain to do a complete source tree reformatting as you'd use for Git repositories?

**Comment by ivanov** [comment]
> Is this just for convenience so you can use the same toolchain to do a complete source tree reformatting as you'd use for Git repositories?

Yes, this, exactly. I'm just thinking of adding 'darker' to my muscle memory (or some alias for it), and it'd be a shame to have to use an alternate command just because it's not in a git folder. You also gain the the ability to see cumulative diff of the choice of black/isort/pylint that you specify, which would take effort without darker.

**Comment by akaihola** [comment]
>it'd be a shame to have to use an alternate command just because it's not in a git folder.

Thinking deeper into this, you're right, it's a valid use case. I'll proceed to review!

File: src/darker/__main__.py (hunk 2 above)
Comment (akaihola) [reply src/darker/__main__.py:87]: An empty `chosen` would indicate that all lines have been deleted.

The fix is going to be merged in #138. Could you wait with this PR and rebase on `master` after that?

Comment (ivanov) [reply src/darker/__main__.py:87]: > An empty chosen would indicate that all lines have been deleted.
Ah, ok, I flipped the "sign" in my head - I thought it was the lines that have changed, so an empty chosen would have been from no changes. Thanks for setting me straight :).

> The fix is going to be merged in #138. Could you wait with this PR and rebase on master after that?

Yep, will do that.

File: src/darker/__main__.py (hunk 1 above)
Comment (akaihola) [reply src/darker/__main__.py:242]: Actually, `git_get_modified_files()` should be called `git_get_modified_python_files()`, and then along the same lines, the new function could be `nogit_get_modified_python_files()`.

File: src/darker/git.py [hunk 3]
```diff
@@ -130,8 +134,13 @@ def _git_check_output_lines(
     """Log command line, run Git, split stdout to lines, exit with 123 on error"""
     logger.debug("[%s]$ %s", cwd, " ".join(cmd))
     try:
-        return check_output(["git"] + cmd, cwd=str(cwd)).decode("utf-8").splitlines()
+        result = run(["git"] + cmd, cwd=str(cwd), check=True, stdout=PIPE, stderr=PIPE)
+        return result.stdout.decode("utf-8").splitlines()
```
Comment (akaihola) [inline src/darker/git.py:138]: Why this change?

File: src/darker/git.py [hunk 4]
```diff
@@ -130,8 +134,13 @@ def _git_check_output_lines(
...
     except CalledProcessError as exc_info:
+        if exc_info.returncode == 129 and exc_info.stderr.decode("utf-8").startswith(
+            "Not a git repository"
+        ):
+            raise NotGitRespository(str(cwd) + " is not a git repository") from exc_info
```
Comment (akaihola) [inline src/darker/git.py:143]: For consistency, I'd use an f-string here like in other corresponding places:
```suggestion
            raise NotGitRespository(f"{cwd} is not a git repository") from exc_info
```

File: src/darker/__main__.py [hunk 5]
```diff
@@ -219,7 +230,16 @@ def main(argv: List[str] = None) -> int:
...
     failures_on_modified_lines = False
     revrange = RevisionRange.parse(args.revision)
-    changed_files = git_get_modified_files(paths, revrange, git_root)
+    try:
+        changed_files = git_get_modified_files(paths, revrange, git_root)
+    except NotGitRespository:
+        changed_files = set()
+        for path in paths:
+            if str(path).endswith(".py"):
+                changed_files.add(Path(path))
```
Comment (akaihola) [inline src/darker/__main__.py:239]: Since `path` is guaranteed to be a `Path`, this could be simplified to
```suggestion
            if path.suffix == ".py":
                changed_files.add(path)
```

File: src/darker/__main__.py (hunk 1 above)
Comment (akaihola) [inline src/darker/__main__.py:242]: This could be simplified too:
```suggestion
                changed_files.update(path.glob("**/*.py"))
```

**Review by akaihola — CHANGES_REQUESTED** [review]

**Review by akaihola — CHANGES_REQUESTED** [review]
We also need unit test cases for non-Git-repository situations, as well as an end-to-end test similar to `darker.tests.test_main.test_main()`.

**Comment by akaihola** [comment]
Hi @ivanov,

I wanted to work on some of my suggestions above, and the only way I could find to do that was to copy this branch to the upstream repository as [akaihola:non-git-repos](https://github.com/akaihola/darker/tree/non-git-repos). 

Do you know if it's possible to switch the branch for this merge request from `ivanov:non-git-repos` to `akaihola:non-git-repos`, or is there a better workflow for doing this kind of a takeover?

Or would you prefer to reset `ivanov:non-git-repos` to point to the latest commit in `akaihola:non-git-repos`?

**Comment by akaihola** [comment]
Ping @ivanov,

I'm preparing a Darker 1.3.0 release. Let me know if you'd like to get this patch in and can spend a few moments on my questions above.

Cheers,
Antti

**Comment by akaihola** [comment]
Ping @ivanov,

I still haven't released 1.3.0. Would you like to get this feature in for the release?

**Comment by akaihola** [comment]
Postponing this feature to at least version 1.3.1.

File: src/darker/git.py [hunk 6]
```diff
@@ -149,23 +153,32 @@ def _git_check_output_lines(
...
-        ).splitlines()
+        )
+        return result.stdout.splitlines()
     except CalledProcessError as exc_info:
+        retval = exc_info.returncode
+        msg = exc_info.stderr
+        if (retval == 129 and msg.startswith("Not a git repository")) or (
+            retval == 1 and msg.startswith("error: could not access ")
+        ):
+            raise NotGitRespository(f"{cwd} is not a git repository") from exc_info
```
Comment (akaihola) [inline src/darker/git.py:172]: I wonder if this is a good way to detect non-Git-directories. The `could not access 'HEAD'` message came up when tests were run on GitHub, but I can't reproduce it locally. We don't know whether that error might occur within Git repositories under different conditions, and this might mask such situations.

Comment (akaihola) [reply src/darker/git.py:172]: ...or could there be an alternative explanation why Git would terminate with `error: Could not access 'HEAD'` than a different Git version?

Pytest temporary directories being created in a directory tree which is a Git repository? Sounds unlikely to me.

Comment (akaihola) [reply src/darker/git.py:172]: No, this is indeed most probably a difference in Git.

In a Debian 11 Docker image:
```
# git --version
git version 2.30.2
# git diff --name-only --relative HEAD -- .
error: Could not access 'HEAD'
```

Same on Debian Buster:
```
# git --version
git version 2.20.1
root@6c76cb727ad0:/# git diff --name-only --relative HEAD -- .
Not a git repository
To compare two paths outside a working tree:
usage: git diff [--no-index] <path> <path>
```

Same on Fedora 30
```
# git --version
git version 2.21.3
# git diff --name-only --relative HEAD -- .
Not a git repository
To compare two paths outside a working tree:
usage: git diff [--no-index] <path> <path>
```

Comment (akaihola) [reply src/darker/git.py:172]: For a non-existing revision, I get this on Git 2.30.2:
```
# git diff --name-only --relative foo -- .
fatal: bad revision 'foo'
```

So at least that won't be confused with a non-Git-directory.

Comment (akaihola) [reply src/darker/git.py:172]: Maybe this would be a more reliable detection:

Git 2.21.3 and 2.30.2:
```
# git rev-parse --is-inside-work-tree
true
```

Git 2.21.3:
```
# git rev-parse --is-inside-work-tree
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

Git 2.30.2:
```
# git rev-parse --is-inside-work-tree
fatal: not a git repository (or any of the parent directories): .git
```

Comment (akaihola) [reply src/darker/git.py:172]: Oops, there really is something funky going on in the GitHub Actions environment!

In `test_main_in_plain_directory(tmp_path, ...)`, when `main()` calls `git_is_repository()`, it gets `True` as the result even though we should be in a clean temporary directory!

See [this build](https://github.com/akaihola/darker/pull/140/checks?check_run_id=3645005237) for details.

Comment (akaihola) [reply src/darker/git.py]: **HEISENBUG!**

I added this test in `test_git.py`:
```python
def test_tmp_path_sanity(tmp_path):
    """Make sure Pytest temporary directories aren't inside a Git repository"""
    try:
        result = git._git_check_output_lines(
            ["rev-parse", "--absolute-git-dir"], tmp_path, exit_on_error=False
        )
    except CalledProcessError as exc_info:
        if exc_info.returncode != 128 or not exc_info.stderr.startswith(
            "fatal: not a git repository"
        ):
            raise
    else:
        output = "\n".join(result)
        raise AssertionError(
            f"Temporary directory {tmp_path} for tests is not clean."
            f" There is a Git directory in {output}"
        )
```

And now all tests pass! How did this change temporary directories in other tests?!?

**Comment by akaihola** [comment]
@ivanov, I'm going to put this feature maybe already in version 1.3.1. Would you still like to review my changes?
